[css-scroll-snap] only preserve a target if we have targets for both axes

nmoucht opened the pull request:
#### 501296a3e648f26c2612145395269563725fed2b
<pre>
[css-scroll-snap] only preserve a target if we have targets for both axes
<a href="https://bugs.webkit.org/show_bug.cgi?id=245989">https://bugs.webkit.org/show_bug.cgi?id=245989</a>
&lt;rdar://100737724&gt;

Reviewed by Simon Fraser.

The preserving of a particular snap target was intended to follow this line
of the spec: &quot;If multiple boxes were snapped before and their snap positions
no longer coincide, then if one of them is focused or targeted, the scroll
container must re-snap to that one and otherwise which one to re-snap to is
UA-defined&quot;. Since it is specified that this is for the case of mulitple boxes
being snapped, change the logic so that we only preserve a target if we are
snapped to multiple boxes.

* LayoutTests/TestExpectations:
* Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
(WebCore::ScrollSnapAnimatorState::setFocusedElementForAxis):
(WebCore::ScrollSnapAnimatorState::preserveCurrentTargetForAxis):
(WebCore::ScrollSnapAnimatorState::resnapAfterLayout):

Canonical link: <a href="https://commits.webkit.org/255493@main">https://commits.webkit.org/255493@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c9e1e4819f81ac888ee14a2d3bc7ef912d978846

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92600 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1819 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23185 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102333 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96601 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1817 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30178 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84999 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98496 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98263 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1226 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79096 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28154 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83144 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82842 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71245 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/83969 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36587 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16769 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/79011 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34374 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17949 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27396 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3810 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38244 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40561 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/81631 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40153 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37107 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18463 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -148,12 +151,9 @@ bool ScrollSnapAnimatorState::resnapAfterLayout(ScrollOffset scrollOffset, const
-        if (*activeHorizontalID != snapOffsetsHorizontal[*activeHorizontalIndex].snapTargetID)
+    // If we have active targets, see if we need to preserve one of them
+    if (activeHorizontalID && activeHorizontalIndex && activeVerticalID && activeVerticalIndex && *activeHorizontalID != snapOffsetsHorizontal[*activeHorizontalIndex].snapTargetID && *activeVerticalID != snapOffsetsVertical[*activeVerticalIndex].snapTargetID)
             snapPointChanged |= preserveCurrentTargetForAxis(ScrollEventAxis::Horizontal);
```
Indent on this line is wrong.

smfr commented on Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
```diff
@@ -148,12 +151,9 @@ bool ScrollSnapAnimatorState::resnapAfterLayout(ScrollOffset scrollOffset, const
-    if (activeHorizontalID && activeHorizontalIndex) {
-        if (*activeHorizontalID != snapOffsetsHorizontal[*activeHorizontalIndex].snapTargetID)
+    // If we have active targets, see if we need to preserve one of them
+    if (activeHorizontalID && activeHorizontalIndex && activeVerticalID && activeVerticalIndex && *activeHorizontalID != snapOffsetsHorizontal[*activeHorizontalIndex].snapTargetID && *activeVerticalID != snapOffsetsVertical[*activeVerticalIndex].snapTargetID)
```
Consider wrapping this long line, or moving the logic into a lambda.

nmoucht committed:
[css-scroll-snap] only preserve a target if we have targets for both axes
https://bugs.webkit.org/show_bug.cgi?id=245989
<rdar://100737724>

Reviewed by Simon Fraser.

The preserving of a particular snap target was intended to follow this line
of the spec: "If multiple boxes were snapped before and their snap positions
no longer coincide, then if one of them is focused or targeted, the scroll
container must re-snap to that one and otherwise which one to re-snap to is
UA-defined". Since it is specified that this is for the case of mulitple boxes
being snapped, change the logic so that we only preserve a target if we are
snapped to multiple boxes.

* LayoutTests/TestExpectations:
* Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
(WebCore::ScrollSnapAnimatorState::setFocusedElementForAxis):
(WebCore::ScrollSnapAnimatorState::preserveCurrentTargetForAxis):
(WebCore::ScrollSnapAnimatorState::resnapAfterLayout):

Canonical link: https://commits.webkit.org/255493@main
